Enhancement: Implementing all the tickets for LBP

Zitzak opened the pull request:
## What did I do
- Rename wrapper to manager in all functions and files.
- Rename `onlyOwner` terminology to `onlyAdmin`.
- Remove transferring funds and joining the pool from the bundled transaction `deployLBPUsingManager` inside `LBPManagerFactory.sol`.
- `SwapFeePercentage` & `primeDaoFeePercentage` now gets passed every time from the frontend, instead of being stored in the `LBPManagerFactory`
- Add function `addLiquidity()` to the LBPManager.
- Add function `removeLiquidity()` to the LBPManager.
- Add functionality to subtract the `primeDaoFeePercentage` from the `projectToken`.
- Add function `setSwapEnabled()` function to the `LBPManager`.
- Add function `removePoolTokens`.
- Adapted the tests accordingly.
- 

_See Issue for more information_

closes #112
closes #111 
closes #116
closes #117 
closes #131

codebuster22 commented on contracts/lbp/LBPWrapper.sol:
```diff
@@ -22,10 +21,12 @@ import "../utils/interface/ILBP.sol";
     address public lbp;
-    uint256 public swapFeePercentage;
-    bool isInitialized;
+    bool public poolFunded;
```
we can rename it to `isPoolFunded`

codebuster22 commented on contracts/lbp/LBPWrapper.sol:
```diff
@@ -44,55 +45,37 @@ contract LBPWrapper {
-     * @param _admin               address of admin/owner of LBP
-     * @param _userData            extra data required by LBP
      */
     function deployLbpFromFactory(
```
let's merge this function and `initialize` into one, and call it `initialize`

Zitzak replied:
Maybe call it something else because we're not really only initializing something, we're actually creating something, the LBP. Why not just `deployLbp`?

Zitzak replied:
Also we're calling `updateWeightGradually` as well.

codebuster22 replied:
Let's call it `initializeLBP`

codebuster22 commented on test/unit/lbp-wrapper.js:
```diff
@@ -100,32 +84,30 @@ describe("Contract: LBPWrapper", async () => {
-          initUserData
+          END_WEIGHTS
         );
+      const receipt = await tx.wait();
```
where are we using this variable `receipt`?

codebuster22 commented on contracts/lbp/LBPWrapper.sol:
we should allow them to enter the receiver address, we don't want to have the tokens with the contract, as of now. Because, we don't have any way to handle those tokens.

Zitzak replied:
I thought the idea is to have the wrapper as owner of the tokens, the admin as owner of the wrapper. And as soon as the admin retrieves his tokens, we will get a percentage, once we start working with fees

codebuster22 replied:
Yes, but the fees we just collected was before it was even successfull. So for now, we will leave it and send tokens to admin. Once we decide something on fees from LBP workshop, we'll implement that function.

codebuster22 commented on contracts/lbp/LBPWrapper.sol:
we can rename this to `joinPoolUsingWrapper` or `joinPool` can also work.

codebuster22 commented on test/unit/lbp-wrapper.js:
```diff
@@ -100,32 +84,30 @@ describe("Contract: LBPWrapper", async () => {
+          END_WEIGHTS
         );
+      const receipt = await tx.wait();
+      setup.Lbp = setup.Lbp.attach(await setup.lbpWrapper.lbp());
```
here, `setup.LBP` is being overwritten. Is this what we wanted?

Zitzak replied:
My understanding was that we attach the instance of LBP that has just been successful created to `setup.Lbp`

codebuster22 replied:
Nope, it returns an instance of connected contract instance.

fabianschu commented on contracts/lbp/LBPWrapper.sol:
```diff
@@ -41,54 +44,34 @@ contract LBPWrapper {
         IERC20[] memory _tokens,
-        uint256[] memory _amounts,
         uint256[] memory _weights,
         bool _swapEnabledOnStart,
```
For my understanding: So `initializeLBP` creates an LBP, but the LBP does not yet contain any tokens, correct? If this is correct, I'm wondering if we should set `_swapEnabledOnStart` to `false` per default, because my understanding is that if the pool is empty no user can in fact swap. Am I missing something?

fabianschu replied:
We might then set `swapEnabledOnStart` to true in the `joinPool` function, I guess?

Zitzak replied:
Yes so we talked about this as well with @codebuster22  but this was a question we still meant to ask during the workshop. I've written it down but indeed, this is what we thought as well

Zitzak replied:
I think I will just change it like you suggested and change it back if needed after a workshop. This way makes more sense

Zitzak replied:
One of the other issues was to implement a `setPause()` function. I implemented that ticket also because we needed a way to pause/unpause anyway

fabianschu commented on test/unit/lbp-wrapper-factory.js:
```diff
@@ -36,7 +36,7 @@ function sortTokens(tokens) {
 }
 
-describe(">> Contract: WrapperFactory", () => {
+describe.only(">> Contract: LBPWrapperFactory", () => {
```
🙃

fabianschu commented on contracts/lbp/LBPWrapper.sol:
I know that the name `joinPool` is commonly used for regular LP pools, where multiple users can add liquidity, but I feel with just one user being able to "join" a pool (= the owner) and since we're creating another abstraction layer here anyways, we could make things a bit clearer by calling this function `fundPool`, but I guess that's a subjective preference ;)

fabianschu commented on contracts/lbp/LBPWrapper.sol:
For my own understanding: Setting `fromInternalBalance` to true basically reduces gas costs for future users that use the pool for swapping. Is that correct?

codebuster22 replied:
yes, it will save gas cost. But all the users might not want to use internal balance.

fabianschu commented on contracts/lbp/LBPWrapper.sol:
Again, why no curly braces anymore? :)

codebuster22 replied:
won't cause `stack too deep` error.

fabianschu commented on test/unit/lbp-wrapper.js:
```diff
@@ -55,78 +58,48 @@ describe("Contract: LBPWrapper", async () => {
 
       sortedTokens = sortTokens(setup.tokenList);
       tokenAddresses = sortedTokens.map((token) => token.address);
+      JOIN_KIND_INIT = 0;
```
Looks like you're reassigning a constant here, when in fact it is not really a constant.

fabianschu commented on contracts/lbp/LBPWrapper.sol:
I understand this function is called by the LBPWrapperFactory upon deploying. Could we change the function to `external`?

codebuster22 replied:
yes, would be great. many functions might need to be marked as external.

fabianschu commented on contracts/lbp/LBPWrapperFactory.sol:
```diff
@@ -47,62 +44,66 @@ contract WrapperFactory is CloneFactory, Ownable {
             _tokens,
-            _amounts,
             _weights,
             _swapEnabledOnStart,
```
See my comment on `_swapEnabledOnStart` in `LBPWrapper.sol`

codebuster22 commented on contracts/lbp/LBPWrapper.sol:
It won't cause any problem of `stack too deep`. Also, we aren't creating any variable here.

BritikovKI commented on contracts/lbp/LBPWrapper.sol:
```diff
@@ -18,11 +18,13 @@ import "../utils/interface/IVault.sol";
 import "../utils/interface/ILBP.sol";
 
 contract LBPWrapper {
+    uint256 public constant swapFeePercentage = 1e12; // 0.0001% is minimum amount.
```
Hmm, why we've changed variable to const? We need swapFeePercentage to be unchangeable?

codebuster22 replied:
that still needs to be decided. As it will mostly used for fundraise, it was decided to have swap fee = 0. But LBP can have 0.0001% as the minimal swap fee. not less than that.

fabianschu replied:
First I thought that if we cannot set the fee to zero, we might want to add a functionality for the wrapper to claim the fees, but 0.0001% is really not worth the effort, I guess. If a project were to raise 1 million $ via the pool, the accrued fee would be just 1$. So probably we can live with some unclaimable dust.

BritikovKI replied:
Ok, I see, then 0.00001 it is

fabianschu commented on contracts/lbp/LBPWrapper.sol:
hmm, isn't this function called by the EOA owner? If so it should be `public`, not `erxternal`.

Zitzak replied:
That could be true. After your comment on a different functions that could have been `external` I looked up in the docs why en when to use what. Reading these linked ([1](https://ethereum.stackexchange.com/questions/19380/external-vs-public-best-practices), [2](https://docs.soliditylang.org/en/v0.8.6/contracts.html?highlight=pure#visibility-and-getters)) I concluded that it could be `external` as well because we will never call this function from inside the contract. And seeing that a call to a `public` function is more expensive I went for `external`. Is my thinking correct or do I misunderstand the meaning of internal and external calls?

fabianschu replied:
Public allows anyone to call the function while external restricts EOAs from calling the function. External is cheaper as you mentioned.

Zitzak replied:
So would this case be then `external` or `public`?

fabianschu replied:
Who's going to call this function? My understanding is that it will be the admin (btw I would find `onlyAdmin` as modifier a clearer name, since we always refer to the person who manages the LBP related things an `admin`). If this is true, the function will be called by an EOA, hence it must be `public`. If this function will/should always only be called by another contract it should be `external`.

Zitzak replied:
Ahh oke. I thought if it's `external` it means external of this contract, which would also be an EOA. On the `onlyOwner` issue, the `admin will also be the owner of the wrapper so that's why we call it `onlyOwner`. So he's the admin of the LBP in the `UI` but the owner of the `LBPWrapper` contract

fabianschu replied:
Yea, I get that, but I think since the admin will *always* take over the role of the owner, and since this ownership can never be transferred, I would find it more logical to not introduce another name and rather directly convey the message that certain functions can only be called by the admin.

codebuster22 replied:
no, external can be called by anyone. When we say `external`, it means only invoke when called by anyone outside the current contract instance. It can be contract or EOA.
But for `public`, it can be invoked by anyone. It can be invoked by current contract instance, other contract or EOA.

codebuster22 replied:
but now, we can optimize it once everything is finalized though

fabianschu replied:
Sorry, my bad. You're right @codebuster22

fabianschu commented on contracts/lbp/LBPWrapper.sol:
```suggestion
    function setPaused(bool _isPaused) public onlyOwner {
```

fabianschu commented on contracts/lbp/LBPWrapper.sol:
Cool, I didn't know there exist ternery operators in Solidity :)

Zitzak replied:
Yeah thought it would be good to use here. Also my first time checking if this would be possible in solidity

fabianschu commented on contracts/lbp/LBPWrapper.sol:
```diff
@@ -18,11 +18,14 @@ import "../utils/interface/IVault.sol";
-    bool isInitialized;
+    bool public isPoolFunded;
+    bool public isInitialized;
+    bool public isPaused;
```
Not sure if we actually need this additional state variable. I believe we can get the information on whether the pool is paused or not from [here](https://github.com/balancer-labs/balancer-v2-monorepo/blob/master/pkg/pool-weighted/contracts/smart/LiquidityBootstrappingPool.sol#L137).

Zitzak replied:
I think you mean the `isPaused` variable right? I thought it would be more expensive to do a call to another contract, that's why I decided to have our own variable. I could implement a function instead like `getPausedEnabled()`?

fabianschu replied:
Intuitively I would say that implementing a function to read the status is cheaper than writing and storing additional data. I would follow your suggestion to implement a getter function, call it isPaused() maybe?

BritikovKI reviewed (approved):
Lgtmt

Zitzak commented:
### What did I do?
- Added custom getter function that retrieves the status of the pool directly from LBP.
- Changed names of `isInitialized` and `isFundPool` and `isPaused` to reflect the way it is done in `Seed.sol`

fabianschu commented on contracts/lbp/LBPWrapper.sol:
```diff
@@ -16,13 +16,17 @@ import "openzeppelin-contracts-sol8/token/ERC20/IERC20.sol";
 import "../utils/interface/ILBPFactory.sol";
 import "../utils/interface/IVault.sol";
 import "../utils/interface/ILBP.sol";
+import "hardhat/console.sol"; //
```
🙃

Zitzak replied:
Damn I even put the comments behind it so I wouldn't forget to remove it :)

fabianschu commented on contracts/lbp/LBPWrapper.sol:
```diff
@@ -16,13 +16,17 @@ import "openzeppelin-contracts-sol8/token/ERC20/IERC20.sol";
-    bool isInitialized;
+    bool public poolFunded;
+    bool public initialized;
+    // bool public isPaused;
```
🙃

fabianschu commented on contracts/lbp/LBPWrapper.sol:
```suggestion
        return !lbp.getSwapEnabled();
```

Zitzak replied:
Easy solution

codebuster22 commented on contracts/lbp/LBPWrapper.sol:
```diff
@@ -18,11 +18,13 @@ import "../utils/interface/IVault.sol";
 import "../utils/interface/ILBP.sol";
 
 contract LBPWrapper {
+    uint256 public constant SWAP_FEE_PERCENTAGE = 1e12; // 0.0001% is minimum amount.
```
comment-> ```// 0.0001% is minimum swap fee allowed by LBP```.

codebuster22 replied:
also, let's add this to WrapperFactory, because someday, we might want to change it as mention by Luuk on latest workshop. LBPWrapperFactory will pass the swap fee to LBPWrapper.

fabianschu commented on contracts/lbp/LBPWrapper.sol:
```suggestion
        lbp.setSwapEnabled(!_isPaused);
```

And remove previous line?

codebuster22 replied:
+1

codebuster22 commented on contracts/lbp/LBPWrapper.sol:
`setSwapEnabled` -> `swapEnabled`, `setSwapEnabled` is a function

Zitzak replied:
I think it should have been `swapEnabledOnStart` looking at the LBP Factory from Balancer

codebuster22 replied:
swapEnabledOnStart is the parameter for the function, the state of LBP have `swapEnabled`

codebuster22 commented on test/unit/lbp-wrapper-factory.js:
```diff
@@ -69,89 +66,94 @@ describe(">> Contract: WrapperFactory", () => {
   });
   context("» set MasterCopy of LBPWrapper", () => {
     it("$ reverts on zero address", async () => {
       await expectRevert(
```
can we use same way of testing, below we have used `expect().to.be.revertedWith()`, let stick to one.

Zitzak replied:
I'm changing it but this is actually something you have written :D

codebuster22 replied:
hm, I don't remember tbh :D

Zitzak commented on contracts/lbp/LBPWrapper.sol:
Yes this is exactly what Fabi also suggested.

codebuster22 replied:
can ignore this, @fabianschu have already mentioned this.

Zitzak commented:
### What did I do
- for `swapFeePercentage`
  - variable will now be set in the constructor of the  `LBPWrapperFactory`
  - added setter function for `swapFeePercentage` to the LBPWrapperFactory`
  - added `getSwapFeePercentage` getter to `LBPWrapper`
- implemented rest of the comments

codebuster22 reviewed (dismissed):
lgtm

Zitzak commented:
We decided to change `onlyOwner` functions to `onlyAdmin` as @fabianschu  suggested. This still has to be changed

codebuster22 commented on contracts/lbp/LBPWrapper.sol:
What if the admin wants to fund using internal balance, we should add an if condition, to check for possible revert.

codebuster22 commented on contracts/lbp/LBPWrapper.sol:
```diff
@@ -16,121 +16,144 @@ import "openzeppelin-contracts-sol8/token/ERC20/IERC20.sol";
-    modifier onlyOwner() {
-        require(msg.sender == owner, "LBPWrapper: only owner function");
+    address public admin;
+    address public primeDaoAddress;
```
rename it to 
```suggestion
    address public beneficiary;
```

codebuster22 commented on contracts/lbp/LBPWrapper.sol:
```diff
@@ -16,121 +16,144 @@ import "openzeppelin-contracts-sol8/token/ERC20/IERC20.sol";
-        address _admin,
+        uint256 _swapFeePercentage,
+        uint256 _primeDaoFeePercentage,
+        address _primeDaoAddress
```
```suggestion
        address _beneficiary
```

codebuster22 commented on contracts/lbp/LBPWrapper.sol:
```diff
@@ -16,121 +16,144 @@ import "openzeppelin-contracts-sol8/token/ERC20/IERC20.sol";
+     * @param _endTime                  end time
+     * @param _endWeights               array of end weights for respective tokens
+     * @param _primeDaoFeePercentage    fee percentage for providing the LBP service
+     * @param _primeDaoAddress          address who is the receiver of the _primeDaoFeePercentage
```
```suggestion
     * @param __beneficiary          address who is the receiver of the _primeDaoFeePercentage
```

fabianschu commented on test/helpers/tokens.js:
```diff
@@ -30,6 +30,14 @@ const getErc20TokenInstances = async (params, from) => {
     })
   );
 
+  if (tokenList.length == 2) {
+    if (tokenList[0].address > tokenList[1].address) {
+      const temp = tokenList[0];
+      tokenList[0] = tokenList[1];
+      tokenList[1] = temp;
+    }
+  }
```
What's this doing?

Zitzak replied:
Balancer wants to have the tokenlist sorted to address. This function does that

codebuster22 commented on contracts/lbp/LBPWrapper.sol:
hm, Until I remember, it would be upfront fee. That would be percentage, and we charge it at the time of creation.

Zitzak replied:
I'm pretty sure it will be a percentage as well. Also from the mocks in the workshops. We did decide on it being 0% to start with

codebuster22 commented on contracts/lbp/LBPWrapper.sol:
```diff
@@ -16,121 +16,182 @@ import "openzeppelin-contracts-sol8/token/ERC20/IERC20.sol";
-            _newOwner != address(0),
-            "LBPWrapper: new owner cannot be zero"
+            _newAdmin != address(0),
+            "LBPWrapper: new admin cannot be zero"
```
I think, we should have a code and in documentation, we can display full error message. Just how balancer, Gnosis does.

fabianschu replied:
I agree that this would be a legit approach. The other one would be to squeeze our error messages below the 32 char limit if possible. Maybe we should discuss w/ the team?

Zitzak commented on contracts/lbp/LBPWrapper.sol:
```diff
@@ -16,121 +16,182 @@ import "openzeppelin-contracts-sol8/token/ERC20/IERC20.sol";
      */
-    function initialize(address _LBPFactory, uint256 _swapFeePercentage)
-        public
+    function retrieveProjectAndFundingToken(address _receiver)
```
This got removed

Zitzak commented on contracts/lbp/LBPWrapper.sol:
```diff
@@ -16,121 +16,182 @@ import "openzeppelin-contracts-sol8/token/ERC20/IERC20.sol";
 import "../utils/interface/ILBPFactory.sol";
 import "../utils/interface/IVault.sol";
 import "../utils/interface/ILBP.sol";
+import "hardhat/console.sol"; // <<<<<<<<<<<<<<< Remove
```
This time I didn't forget actually :D the PR is just still open while I'm pushing work

fabianschu commented on contracts/lbp/LBPManager.sol:
```diff
@@ -112,52 +126,60 @@ contract LBPManager {
-            fromInternalBalance: _fromInternalBalance,
-            assets: _tokens
+            fromInternalBalance: false, // It is not possible to add liquidity thought the internal Vault balance.
+            assets: _tokenList
```
In light of what we talked about yesterday @Zitzak: this is under the assumption that we already receive the parameters sorted by address numerically from the frontend, right?

fabianschu commented on contracts/lbp/LBPManager.sol:
```diff
@@ -51,25 +55,35 @@ contract LBPManager {
+     * @param _beneficiary              The address that receives the _primeDaoFeePercentage.
+     * @param _name                     Name of the LBP.
+     * @param _symbol                   Symbol of the LBP.
+     * @param _tokenList                Sorted array containing two parameters:
```
numerically sorted ascending is correct, right?

```suggestion
     * @param _tokenList                Numerically sorted array (ascending) containing two addresses:
```

Zitzak replied:
Yes, and I'll add you're suggestion

fabianschu commented on contracts/lbp/LBPManager.sol:
```diff
@@ -112,52 +126,60 @@ contract LBPManager {
-     * @param _fromInternalBalance      fund tokens from the internal user balance
-     * @param _userData                 userData specifies the type of join
+     * @dev                             Subtracts the primeDaoFee and adds liquidity to the LBP.
+     * @param _tokenList                Sorted array containing two parameters:
```
```suggestion
     * @param _tokenList                Numerically sorted array (ascending) containing two addresses:
```

fabianschu commented on contracts/lbp/LBPManager.sol:
```diff
@@ -233,28 +255,36 @@ contract LBPManager {
      */
-    function feeAmountRequired() internal view returns (uint256 feeAmount) {
-        feeAmount = (amounts[0] * primeDaoFeePercentage) / HUNDRED_PERCENT;
+    function feeAmountRequired(uint8 _fundingTokenIndex)
```
I would follow the convention and let internal function be preceded with an underscore (or does solhint complain then)?

```suggestion
    function _feeAmountRequired(uint8 _fundingTokenIndex)
```

Zitzak replied:
No I didn't think about the convention. Changed it

fabianschu commented on test/helpers/tokens.js:
```diff
@@ -32,9 +32,7 @@ const getErc20TokenInstances = async (params, from) => {
-      const temp = tokenList[0];
-      tokenList[0] = tokenList[1];
-      tokenList[1] = temp;
+      tokenList.reverse();
```
nice! never heard of that method

Zitzak commented:
Test Coverage is 100%

Zitzak commented on contracts/lbp/LBPManager.sol:
```diff
@@ -112,52 +126,60 @@ contract LBPManager {
+                                            - The address of the funding token being exchanged for the project token.
+     * @param _fundingTokenIndex        Index for the _tokenList array for the funding token.
+     * @param _sender                   Address of the liquidity provider.
+     * @param _userData                 UserData object that specifies the type of join.
```
You're right, we have to make sure they know. I've written it down, in case Mihir hasn't already put it into the specs.

codebuster22 replied:
Yes, a handover document will be created, which will ensure this parameters and how to build them.

BritikovKI commented on contracts/lbp/LBPManager.sol:
```diff
@@ -0,0 +1,290 @@
+        uint256[] memory _startTimeEndTime,
+        uint256[] memory _endWeights,
+        uint256 _swapFeePercentage,
+        uint256 _primeDaoFeePercentage
```
Hmm, I am not clear here, as I understand we need to set fee to 0(according to issue #112 ), but here the fee is configurable? Which exact fee do we need to set to 0?

Zitzak replied:
We kept track of all the changes that got decided on in the Miro board as well as a spec doc. I did not change the issue to reflect the constant changes and when creating this final PR, I just added all the issues that were related to this not thinking it through that the reviewer would, of course, use the issues as reference.

In the case of the `swapFeePercentage`, Balancer did not allow zero percentage so we first used the lowest possible, set it as a constant in the `LBPManagerFactory`, passing it to the manager on deploying it. We ended up with a decision where a swap fee should be used and passed from the dapp every time a LBP gets created through the wizard. That is the implementation you see here.

I updated the issues to reflect the changes. My apologies for it being a bit messy in this way.

BritikovKI replied:
Hmm, so `_primeDaoFeePercentage` should not be changed after the contract was deployed?

Zitzak replied:
Yes indeed, the fee for the service will be set in the wizard when creating the LBP. This fee will stay the same for that LBP and LBPManager instance

BritikovKI commented on contracts/lbp/LBPManager.sol:
```diff
@@ -0,0 +1,290 @@
+        uint256[] memory _weights,
+        uint256[] memory _startTimeEndTime,
+        uint256[] memory _endWeights,
+        uint256 _swapFeePercentage,
```
Don't know where to place it so I'll place it here, according to the #117 We need to have a fee which is **changeable** but there is no `SetFee` function or smth. Have this functionality been implemented, or have I misunderstood the issue?

Zitzak replied:
See explanation above

BritikovKI commented on contracts/lbp/LBPManager.sol:
```diff
@@ -0,0 +1,290 @@
+        address payable _receiver,
+        bool _toInternalBalance,
+        bytes memory _userData
+    ) external {
```
It should be `OnlyAdmin`, no?

Zitzak replied:
You're right, good catch

BritikovKI commented on contracts/lbp/LBPManager.sol:
```diff
@@ -0,0 +1,290 @@
+     * @param _toInternalBalance        Balancer's Vault option to send the liquidity from the LBP to the _receivers internal Vault balance.
+     * @param _userData                 UserData object that specifies the type of exit.
+     */
+    function removeLiquidity(
```
If we add a new function, it should be always tested in the same PR, as I understand there are no tests rn?

BritikovKI replied:
Yep, there are, missed tests(test file wasn't displayed because of +1000 loc)

BritikovKI commented on contracts/lbp/LBPManager.sol:
```diff
@@ -0,0 +1,290 @@
+     * @dev                             Can pause/unpause trading.
+     * @param _swapEnabled              Enables/disables swapping.
+     */
+    function setSwapEnabled(bool _swapEnabled) external onlyAdmin {
```
Is this function tested?

BritikovKI replied:
Yep, it is, sorry, missed the tests

BritikovKI commented on contracts/lbp/LBPManager.sol:
```diff
@@ -0,0 +1,290 @@
+*/
+
+// SPDX-License-Identifier: GPL-3.0-or-later
+
```
This PR states that it closes #132 . Can you pls help me here, where can i find states described in #132 ?

Zitzak replied:
This ticket should have been closed earlier. I took what was decided in the workshops by Doug, that we need states, and transferred it to an issue. In the comments of the issue, Mihir commented that these states are needed for the frontend but should not reflect contract states. After that, this issue could have been closed

BritikovKI replied:
Hmm, ok, it is still kinda closed in the description of this PR, can you pls remove it then and close it manually?

Zitzak replied:
Done

Zitzak commented on contracts/lbp/LBPManager.sol:
```diff
@@ -0,0 +1,290 @@
+/*
```
Totally agree, I don't feel like this being effective being such a big PR. It was not at all on purpose but the specs kept changing all the time and I kept implementing them before the PR's got through. Thanks for the feedback

BritikovKI replied:
Well, yep, that's understandable. Usually the best connection is 1 issue - 1 PR....

BritikovKI replied:
So ideally it should be new issue - new branch - new PR

Zitzak replied:
I Will make sure not to let them get so big next time :)

BritikovKI commented on contracts/lbp/LBPManager.sol:
```diff
@@ -0,0 +1,290 @@
+     * @param _userData                 UserData object that specifies the type of exit.
+     */
+    function removeLiquidity(
+        IERC20[] memory _tokenList,
```
Hmm, do we need to check tokenList here, what if we remove Liquidity not from all tokens? Or if we try to removeLiquidity for tokens that were not previously sent?

BritikovKI replied:
P.S. as I understand we can call this function few times?

Zitzak replied:
Balancer will return an error if we try to withdraw with tokens not used before. Also, I think it's not possible to only withdraw one half of the pools liquidity. I do wonder now why we pass the `tokenList` every time and not just take them from the state variables defined by the call `initializeLBP`? That would make it cleaner as well right?

Zitzak replied:
@codebuster22 maybe you have a reason why we shouldn't just use the state variables?

Zitzak replied:
And we can only call this function once because of the require statement that checks if the manager has a balance of pool tokens.

```
        require(
            lbp.balanceOf(address(this)) > 0,
            "LBPManager: manager does not have any pool tokens to remove liquidity"
        );
```

BritikovKI replied:
Yep, I think it will be a lot more clear without passing the list of tokens...
Maybe it is done because of the inheritance reasons, but I don't think that it is needed

BritikovKI replied:
We can call this function only once if we pass the whole list of tokens, but what if we have 4 tokens in the pool and we pass only 2?

Zitzak replied:
Balancer LBP can't have 4 tokens in a pool. So in case of the LBP this would not work. Also I think a LBP with multiple tokens doesn't really make sense

Zitzak replied:
Could you explain what you mean by inheritance reason? That the state variable would be available in inherited contracts so you might want to pass them as arguments?

BritikovKI replied:
Hmm if it can have only 2 tokens, why don't we limit array sizes? I will add an issue to do it

codebuster22 replied:
LBP will only have two tokens. Not more than that. The reason I added _tokenList was, to make testing easier. Once we have all the functionality implemented. We were going to optimize it.
Here we are using `Exact BPT/Pool tokens in for Amounts out`. So we specify how much `BPT Tokens` we want to give, and the minimum amount of `Tokens` we want out. As this PR was already heavy. We wanted to complete the functionality and try to do housekeeping work and optimization in next.

BritikovKI commented on contracts/lbp/LBPManager.sol:
```diff
@@ -0,0 +1,290 @@
+        );
+
+        // To remove all funding from the pool. Initializes to [0, 0]
+        uint256[] memory _minAmountsOut = new uint256[](2);
```
Btw, why the array length here is exactly 2?

Zitzak replied:
Because Balancers LBP only functions with two tokens. But good point, I made the size dynamic

BritikovKI commented on contracts/lbp/LBPManager.sol:
```diff
@@ -0,0 +1,290 @@
+        );
+        require(
+            lbp.balanceOf(address(this)) > 0,
+            "LBPManager: wrapper does not have any pool tokens to remove liquidity"
```
```suggestion
            "LBPManager: manager does not have any pool tokens to remove liquidity"
```

BritikovKI commented on contracts/lbp/LBPManager.sol:
```diff
@@ -0,0 +1,290 @@
+
+        require(
+            lbp.balanceOf(address(this)) > 0,
+            "LBPManager: wrapper does not have any pool tokens to withdraw"
```
```suggestion
            "LBPManager: manager does not have any pool tokens to withdraw"
```

BritikovKI commented on test/test-init.js:
```diff
@@ -38,7 +38,7 @@ const getGnosisProxyInstance = async (setup) => {
 };
 
 const getLBPWrapperFactory = async (setup) => {
```
```suggestion
const getLBPManagerFactory = async (setup) => {
```

BritikovKI commented on test/unit/lbp-manager.js:
```diff
@@ -0,0 +1,1157 @@
+    );
+  });
+
+  describe("# deploy LBP using Wrapper", () => {
```
```suggestion
  describe("# deploy LBP using Manager", () => {
```

BritikovKI commented on test/unit/lbp-manager.js:
```diff
@@ -0,0 +1,1157 @@
+  });
+
+  describe("# deploy LBP using Wrapper", () => {
+    describe("$ deploy LBP using Wrapper fails", () => {
```
```suggestion
    describe("$ deploy LBP using Manager fails", () => {
```

BritikovKI commented on test/unit/lbp-manager.js:
```diff
@@ -0,0 +1,1157 @@
+        );
+      });
+    });
+    describe("$ deploy LBP using Wrapper succeeds", () => {
```
```suggestion
    describe("$ deploy LBP using Manager succeeds", () => {
```

BritikovKI commented on test/unit/lbp-manager.js:
```diff
@@ -0,0 +1,1157 @@
+        await expect(
+          lbpManagerInstance.connect(admin).withdrawPoolTokens(admin.address)
+        ).to.be.revertedWith(
+          "LBPManager: wrapper does not have any pool tokens to withdraw"
```
```suggestion
          "LBPManager: manager does not have any pool tokens to withdraw"
```

BritikovKI commented on test/unit/lbp-manager.js:
```diff
@@ -0,0 +1,1157 @@
+            .connect(admin)
+            .removeLiquidity(tokenAddresses, admin.address, false, exitUserData)
+        ).to.be.revertedWith(
+          "LBPManager: wrapper does not have any pool tokens to remove liquidity"
```
```suggestion
          "LBPManager: manager does not have any pool tokens to remove liquidity"
```

codebuster22 reviewed (approved):
lgtm

kinspotting commented on contracts/lbp/LBPManagerFactory.sol:
```diff
@@ -0,0 +1,141 @@
+     * @param _LBPFactory               The address of Balancers LBP factory.
+     */
+    function setLBPFactory(address _LBPFactory) external onlyOwner {
+        require(
```
logic of require statements on line 53, 57, 68 and 72 is the same, maybe it makes sense to create a generic modifier for it

kinspotting commented on contracts/lbp/LBPManagerFactory.sol:
```diff
@@ -0,0 +1,141 @@
+ * @dev   Enable PrimeDAO governance to create new LBPManager contracts.
+ */
+contract LBPManagerFactory is CloneFactory, Ownable {
+    address public lbpManagerMasterCopy;
```
why do we call this variable `lbpManagerMasterCopy` do we have any other master copies?

Zitzak replied:
If I understand it correctly then you mean why not call it just `masterCopy`?

codebuster22 replied:
yes

kinspotting replied:
that's correct :)

kinspotting commented on contracts/lbp/LBPManagerFactory.sol:
```diff
@@ -0,0 +1,141 @@
+     * @param _swapFeePercentage        Percentage of fee paid for every swap in the LBP.
+     * @param _primeDaoFeePercentage    Percentage of fee paid to PrimeDao for providing the service of the LBP Manager.
+     */
+    function deployLBPUsingManager(
```
why do we have to specify that the deployment happens with `Manager`, can we just name it `deployLBP`?

Zitzak replied:
In a later PR, I changed it to `deployLBPManger`. Would that also be okay or change it to `deployLBP`?

kinspotting reviewed (changes requested):
left a review

fabianschu commented on contracts/lbp/LBPManager.sol:
```diff
@@ -0,0 +1,290 @@
+     */
+    function addLiquidity(
+        IERC20[] memory _tokenList,
+        uint8 _projectTokenIndex,
```
Afaiu, the only way to check that in a bulletproof manner onchain is to compare against the addresses of funding tokens, which we know and which are fixed (?) But maybe it would be sufficient to check that the weight of the alleged project token is smaller than the weight of the alleged funding token? Anyways, I'm curious why you think we need to implement additional safeguards against faulty inputs from the UI here :)

Zitzak replied:
So we had a call with Doug on Friday. He said it would be better to sort it onchain because there're only two tokens, which should not be costly but just a direct comparison `>`. This change will be implemented in a different PR. I already started on it but encountered an issue about which I've sent a question to Doug.

Zitzak replied:
So we will probably end up with the project and funding token always coming in a certain order, and a quick comparison to sort the array

kinspotting replied:
> So we will probably end up with the project and funding token always coming in a certain order, and a quick comparison to sort the array

sounds good to me

Zitzak commented on contracts/lbp/LBPManager.sol:
```diff
@@ -0,0 +1,290 @@
+
+/**
+ * @title PrimeDAO LBPManager contract
+ * @dev   Smart contract for managing a Balancer LBP.
```
How would this be?
```
/**
 * @title LBPManager contract
 * @dev   Smart contract for managing interactions with a Balancer LBP.
 */
```

I removed the PrimeDao reference because of a comment from Doug in last weeks call where he said it would be nice to remove the reference in all the contract to make it more generic

codebuster22 reviewed (approved):
Lgtm

kinspotting reviewed (approved):
lgtm
